lib/block-supports/duotone.php

```diff
@@ -0,0 +1,97 @@
+<?php
+/**
+ * Duotone block support flag.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Registers the style and colors block attributes for block types that support it.
+ *
+ * @param WP_Block_Type $block_type Block Type.
+ */
+function gutenberg_register_duotone_support( $block_type ) {
+	$has_duotone_support = false;
+	if ( property_exists( $block_type, 'supports' ) ) {
+		$has_duotone_support = _wp_array_get( $block_type->supports, array( 'color', 'duotone' ), false );
+	}
+
+	if ( $has_duotone_support ) {
+		if ( ! $block_type->attributes ) {
+			$block_type->attributes = array();
+		}
+
+		if ( ! array_key_exists( 'style', $block_type->attributes ) ) {
+			$block_type->attributes['style'] = array(
+				'type' => 'object',
+			);
+		}
+	}
+}
+
+/**
+ * Render out the duotone stylesheet and SVG.
+ *
+ * @param  WP_Block_Type $block_type       Block type.
+ * @param  array         $block_attributes Block attributes.
+ * @param  string        $block_content    Rendered block content.
+ *
+ * @return string filtered block content.
+ */
+function gutenberg_render_duotone_support( $block_content, $block ) {
+	$block_type = WP_Block_Type_Registry::get_instance()->get_registered( $block['blockName'] );
+
+	$duotone_support = false;
+	if ( $block_type && property_exists( $block_type, 'supports' ) ) {
+		$duotone_support = _wp_array_get( $block_type->supports, array( 'color', 'duotone' ), false );
+	}
+
+	$has_duotone_attribute = isset( $block['attrs']['style']['color']['duotone'] );
+
+	if (
+		! $duotone_support ||
+		! $has_duotone_attribute
+	) {
+		return $block_content;
+	}
+
+	$duotone_attribute = $block['attrs']['style']['color']['duotone'];
+
+	$slug_or_id = isset( $duotone_attribute['slug'] ) ? $duotone_attribute['slug'] : uniqid();
+
+	$duotone_values = $duotone_attribute['values'];
+	$duotone_id     = 'wp-duotone-filter-' . $slug_or_id;
+
+	$block_class = gutenberg_get_block_default_classname( $block_type->name );
```

Comment:
What happens if the block types doesn't support it's on default className? Maybe we can avoid using $block_class entirely here and just use the duotone id as a selector?

Reply:
The problem that this was supposed to solve was not accidentally applying duotone to inner blocks that contained markup that matched the selector. At one point I was using just the `img` and `video` selectors for the cover block, so I thought that adding the class would prevent people from accidentally making the mistake of applying the selector to other things inside inner blocks.

Unfortunately this doesn't work for nesting the same kind of block inside itself—like cover inside cover. That's what I was asking about in https://github.com/WordPress/gutenberg/pull/26752#discussion_r600754335

Reply:
yeah, in these cases, if we allow selectors, they need to be precise `> something > something`

Reply:
I see the "selector" as a secondary use case, ideally we avoid it entirely and `$block_class` is not as simple as you just did. Some block types opt-out of it, others do provide their own classes. Some blocks don't have any (paragraph, headings for instance), so for me, this is just useless complexity.

Reply:
Does it look like we could benefit from having a `gutenberg_get_block_selector` function? We already do that at https://github.com/WordPress/gutenberg/blob/trunk/lib/class-wp-theme-json.php#L444 and it could be extracted to a separate function to be reused here. Otherwise, this will fail for blocks such as paragraph, heading, etc.